app/questionnaire/questionnaire_schema.py

```diff
@@ -215,20 +216,52 @@ def get_post_submission(self) -> ImmutableDict:
         schema: ImmutableDict = self.json.get("post_submission", ImmutableDict({}))
         return schema
 
+    def _is_list_name_in_rule(
+        self, when_rules: Union[dict, Sequence], list_name: str
+    ) -> bool:
+        rules: Union[dict, Sequence]
+        if isinstance(when_rules, dict) and any(
+            operator in when_rules for operator in OPERATION_MAPPING
+        ):
+            rules = self.get_operand(when_rules)
+        else:
+            rules = when_rules
+
+        if isinstance(rules, Sequence):
+            for rule in rules:
+                # old rules
+                if isinstance(rule, dict) and "list" in rule:
+                    return rule.get("list") == list_name
+
+                # New rules
+                if isinstance(rule, dict) and "source" in rule:
+                    return (
+                        rule.get("source") == "list"
+                        and rule.get("identifier") == list_name
+                    )
+
+                # Nested rules
+                if isinstance(rule, dict) and any(
+                    operator in rule for operator in OPERATION_MAPPING
+                ):
+                    return self._is_list_name_in_rule(rule, list_name)
```

Comment:
I don't think this is instance sequence check is needed?. If it is the old rules, you will have a list of rules, or if it is the new rules, then the operands would be a list as well?.

Also, could continue early and avoid the multiple isisntance check.

```suggestion
        for rule in rules:
            if not isinstance(rule, dict):
                continue

            # old rules
            if "list" in rule:
                return rule.get("list") == list_name

            # New rules
            if "source" in rule:
                return (
                    rule.get("source") == "list"
                    and rule.get("identifier") == list_name
                )

            # Nested rules
            if any(
                operator in rule for operator in OPERATION_MAPPING
            ):
                return self._is_list_name_in_rule(rule, list_name)
```